File: Homework02-golang/Exercise04/payroll.go
/*
 * File: payroll.go
 * Author: Eduardo R Torres
 * Course: COTI 4039-KJ1
 * Date: 10/03/2024
 * Purpose: This program evaluates a group of payroll employees to display their
 *          Info and Calculate their Salary.
 */

package main

import "fmt"

type department int

const (
	Finance department = iota
	HumanResources
	InformationTechnology
	Marketing
	Sales
)

// Returns the string representation of the department.
func (e department) String() string {
	if Finance > e || e > Sales {
		return "unknown department"
	}
	departments := [...]string{"Finance", "Human Resources", "Informartion Technology", "Marketing", "Sales"}
	return departments[e]
}

type employee struct {
	id   string
	name string
	department
}

type pago interface {
	weeklySalary() float64
}

type employeeHour struct {
	employee
	hoursWorked int
	payRate     float64
}

func (e employeeHour) weeklySalary() float64 {
	overTime := e.hoursWorked - 40
	overRate := e.payRate * 1.5
	extraPay := 0.0

	if overTime > 0 {
		extraPay = float64(overTime) * overRate
	}

	return e.payRate*float64(e.hoursWorked) + extraPay
}

type employeeSales struct {
	employee
	salesAmnt      float64
	commissionRate float64
}

func (e employeeSales) weeklySalary() float64 {
	return e.commissionRate * float64(e.salesAmnt)
}

func printData(emp pago) {
	// Cast the pago interface to employeeHour to access employee details
	if empData, ok := emp.(employeeHour); ok {
		fmt.Println()
		fmt.Println("Id:", empData.id)
		fmt.Println("Name:", empData.name)
		fmt.Println("Department:", empData.department)
		fmt.Println("Hours Worked:", empData.hoursWorked)
		fmt.Printf("Pay Rate: %.2f\n", empData.payRate)
		fmt.Printf("Salary: $%.2f\n", empData.weeklySalary())

	} else if empDataS, ok := emp.(employeeSales); ok {
		fmt.Println()
		fmt.Println("Id:", empDataS.id)
		fmt.Println("Name:", empDataS.name)
		fmt.Println("Department:", empDataS.department)
		fmt.Println("Sales Ammount:", empDataS.salesAmnt)
		fmt.Printf("Commission Rate: %.2f\n", empDataS.commissionRate)
		fmt.Printf("Salary: $%.2f\n", empDataS.weeklySalary())

	} else {
		fmt.Printf("Unknown employee type")
		fmt.Println()
	}
}

//  Starts the execution of the program.
func main() {
	pagos := []pago{
		employeeHour{employee: employee{id: "1111", name: "John Doe", department: Marketing}, hoursWorked: 40, payRate: 15.0},
		employeeSales{employee: employee{id: "2222", name: "Johnny Doesn't", department: Sales}, salesAmnt: 5000.00, commissionRate: .15},
		employeeHour{employee: employee{id: "3333", name: "Johnnatan Doever", department: HumanResources}, hoursWorked: 45, payRate: 22.0},
		employeeSales{employee: employee{id: "4444", name: "John Joestar", department: Sales}, salesAmnt: 38, commissionRate: 1.0},
		employeeHour{employee: employee{id: "5555", name: "Johnny Does", department: Finance}, hoursWorked: 38, payRate: 20.0},
		employeeSales{employee: employee{id: "6666", name: "Johnnatan Jostar", department: Sales}, salesAmnt: 38, commissionRate: .10},
	}

	fmt.Println("These are the Employees...")
	for _, pago := range pagos {
		printData(pago)
	}
}

File: Homework05-Prolog/payroll.pl
#!/usr/bin/perl
use strict;
use warnings;

# Define employee data
my @employees = (
    {
        id => '1111',
        name => 'John Doe',
        department => 'Information Technology',
        type => 'hourly',
        hourly_pay_rate => 15.00,
        hours_worked => 40
    },
    {
        id => '2222',
        name => 'Jane Doe',
        department => 'Sales',
        type => 'salesperson',
        commission_rate => 0.15,
        sales_amount => 5000.00
    },
    {
        id => '3333',
        name => 'Alice Smith',
        department => 'Finance',
        type => 'hourly',
        hourly_pay_rate => 20.00,
        hours_worked => 45
    },
    {
        id => '4444',
        name => 'Bob Johnson',
        department => 'Marketing',
        type => 'salesperson',
        commission_rate => 0.10,
        sales_amount => 3000.00
    },
    {
        id => '5555',
        name => 'Charlie Brown',
        department => 'Human Resources',
        type => 'hourly',
        hourly_pay_rate => 18.00,
        hours_worked => 38
    }
);

# Calculate weekly salary
sub calculate_weekly_salary {
    my ($employee) = @_;
    if ($employee->{type} eq 'hourly') {
        my $hours = $employee->{hours_worked};
        my $rate = $employee->{hourly_pay_rate};
        if ($hours > 40) {
            return (40 * $rate) + (($hours - 40) * $rate * 1.5);
        } else {
            return $hours * $rate;
        }
    } elsif ($employee->{type} eq 'salesperson') {
        return $employee->{commission_rate} * $employee->{sales_amount};
    }
    return 0;
}

# Display employee data and weekly salary
foreach my $employee (@employees) {
    my $salary = calculate_weekly_salary($employee);
    print "ID: $employee->{id}\n";
    print "Name: $employee->{name}\n";
    print "Department: $employee->{department}\n";
    print "Weekly Salary: \$" . sprintf("%.2f", $salary) . "\n";
    print "-----------------------------\n";
}